README.md
# EcFinder

EcFinder tries to find an EC number for an enzyme name
using exact and fuzzy search.

It automatically downloads the EC data file from Expasy.

## Installation

It can be installed using ``pip install ec_finder``.

## CLI

The CLI tool ``ec-find`` is provided by the package.

### Setup

The first time the tool is called from the command line, the enzyme
data file is downloaded from Expasy. To update the file if a new file
is available, use ``ec-find update``.
To remove the file (e.g. to force an update), use ``ec-find cleanup``.

### Search

Use ``ec-find "enzyme name"`` to find the EC number for an enzyme name.
Note the quotes, necessary if there are spaces.

Use ``ec-find filename 1`` to find the EC number for all enzyme names
in a file, where each line contains only an enzyme name each.

Use ``ec-find filename colnum`` to find the EC number for all enzyme names
in a Tab-separated file, where enzyme names are in the column with the
specified 1-based column number.

Use ``ec-find filename colnum --separator S`` to find the EC number for all
enzyme names in a file separated by separator ``S``.

## API

The first time that ``ec_finder`` is imported, the package
data is downloaded from Expasy.

The ``ec_finder.update()`` function can be used to check if new
enzyme data is avalaible at Expasy and, if so, download it and update
the data. The ``ec.finder.cleanup()`` function deletes the data.

The ``ec_finder.search(enzyme_name)`` function is used for searching
for an enzyme name.

## Output format

The output format consists of three fields, separated by the separator
indicated in the ``--separator`` option (CLI), or by the ``separator`` argument
of ``search()`` (API). By default the Tab is used.

The three fields are:
- result comment
- result EC
- enzyme name (input)

The result comment may be:
- MAIN, the enzyme name was found as one of the main enzye names.
- ALT, the enzyme name was found as one of the alternative enzyme names.
- FUZZY:foobar, the enzyme name was not found, but a similar one (foobar)
  was found
- NOT\_FOUND, the enzyme name was not found (in this case EC is empty)

If the input of the CLI tool ``ec-find`` is a file, then each file line
is output after the three fields reported above, separated from it by the
separator (default: tab).

## Settings

The minimum score for the fuzzy search is by default "90" and can be set
to a value between 0 and 100 by the option ``--min-score`` of the CLI
or by the argument ``min_score`` of the API function ``search()``.

The separator in the output is by default TAB and can be set by
the CLI option ``--separator`` (which also changes the input separator,
if the input is a file) or by the API function ``search()`` argument
``separator``.


CHANGELOG.txt
=== 0.1 ===
- initial release


tests/test_ec_finder.py
#
# (c) 2023 Giorgio Gonnella
#
import os
import functools
from tempfile import TemporaryDirectory
import ec_finder

TEST_ENZYME_DAT = """\
ID   1.1.1.1
DE   Alcohol dehydrogenase.
AN   ADH.
AN   AHD2.
CA   An NAD+ oxidoreductase.
CF   Zinc.
CC   -!- A zinc protein. Animal and plant ADH's are classified in
CC       etc etc etc
//
ID   1.1.1.2
DE   D/L-glycerol-3-phsphate dehydrogenase.
AN   GLYCEROL-3-PHOSPHATE DEHYDROGENASE (ACCEPTOR).
AN   GLYCEROPHOSPHATE DEHYDROGENASE.
AN   GPDH.
CA   An NAD(P)+ oxidoreductase.
CC   -!- The enzyme is a complex of three polypeptides of which the
CC       etc etc etc
//
ID   1.1.1.3
DE   2-oxoacid:ferredoxin oxidoreductase.
AN   Oxidoreductase, 2-oxoacid:ferredoxin.
AN   2-ketoacid-ferredoxin oxidoreductase.
CA   An NAD(P)+ oxidoreductase.
CF   Iron-sulfur.
CC   -!- A group of enzymes, formerly listed under EC 1.2.7, catalysing
CC       etc etc etc
//
"""

def test_parse_enzyme_dat():
    with TemporaryDirectory() as tmpdir:
        filename = os.path.join(tmpdir, "enzyme.dat")
        with open(filename, "w") as f:
            f.write(TEST_ENZYME_DAT)
        index = ec_finder.ec_finder.parse_enzyme_dat(filename)
        assert len(index["main"]) == 3
        assert len(index["alt"]) == 7
        assert "Alcohol dehydrogenase" in index["main"]
        assert "1.1.1.2" in index["main"].values()
        assert "GLYCEROPHOSPHATE DEHYDROGENASE" in index["alt"]
        assert "1.1.1.2" in index["main"].values()

def test_search():
    with TemporaryDirectory() as tmpdir:
        filename = os.path.join(tmpdir, "enzyme.dat")
        with open(filename, "w") as f:
            f.write(TEST_ENZYME_DAT)
        index = ec_finder.ec_finder.parse_enzyme_dat(filename)
        search = functools.partial(ec_finder.ec_finder.search, index=index)
        # Test exact match with main name
        assert search("Alcohol dehydrogenase") == \
                      "MAIN\t1.1.1.1\tAlcohol dehydrogenase"
        # Test separator argument
        assert search("Alcohol dehydrogenase", separator=';') == \
                      "MAIN;1.1.1.1;Alcohol dehydrogenase"
        # Test exact match with alternative name
        assert search("GLYCEROPHOSPHATE DEHYDROGENASE") == \
                      "ALT\t1.1.1.2\tGLYCEROPHOSPHATE DEHYDROGENASE"
        # Test fuzzy match
        assert search("alcohol dehydrogenase") == \
            "FUZZY:Alcohol dehydrogenase\t1.1.1.1\talcohol dehydrogenase"
        # Test no match
        assert search("alcohol dehydro") == \
            "NOT_FOUND\t\talcohol dehydro"
        # Test min_score argument
        assert search("alcohol dehydro", min_score=20) == \
            "FUZZY:Alcohol dehydrogenase\t1.1.1.1\talcohol dehydro"



ec_finder/ec_finder.py
from pathlib import Path
from shutil import rmtree
from loguru import logger
import sys
import os
import sh
from platformdirs import PlatformDirs
from thefuzz import process

__version__="0.1"
__appname__="EcFinder"
__author__="ggonnella"

ED_BASEFILENAME = "enzyme.dat"
FTP_URL = "ftp.expasy.org/databases/enzyme/" + ED_BASEFILENAME
APPDATADIR = Path(PlatformDirs(__appname__, __author__, \
                              version=__version__).user_data_dir)
ENZYMESDAT = str(APPDATADIR / ED_BASEFILENAME)

def enable_logger(level):
  logger.remove()
  logger.enable("ec_finder")
  msgformat_prefix="<green><dim>{time:YYYY-MM-DD HH:mm:ss}</></>"
  msgformat_content="<level><normal>{level.name}: {message}</></>"
  logger.add(sys.stderr, format=f"{msgformat_prefix} {msgformat_content}",
             level=level)

def download(remotefile, localfile):
  timestampfile = localfile + ".timestamp"
  args = ["-o", localfile, "-w", "%{size_download}", "-R"]
  if os.path.exists(timestampfile):
    args += ["-z", timestampfile]
  try:
    ret = sh.curl("ftp://" + remotefile, *args)
    failed = False
  except sh.ErrorReturnCode:
    failed = True
  if failed:
    ret = sh.curl("https://" + remotefile, *args)
  downloaded = int(ret.rstrip())
  if downloaded > 0:
    sh.touch(timestampfile)
    return True
  else:
    return False

def update(force_download=False):
  enable_logger("INFO")
  logger.info("Data directory: {}".format(APPDATADIR))
  logger.info("Looking for updated enzyme data...")
  if force_download:
    cleanup()
  APPDATADIR.mkdir(parents=True, exist_ok=True)
  updated = download(FTP_URL, ENZYMESDAT)
  if updated:
    logger.info("Updated enzyme data file.")
  else:
    logger.info("No enzyme data file update needed.")

def cleanup():
  enable_logger("INFO")
  logger.info("Data directory: {}".format(APPDATADIR))
  if APPDATADIR.exists():
    for path in Path(APPDATADIR).glob("**/*"):
        if path.is_file():
            path.unlink()
        elif path.is_dir():
            rmtree(path)
  logger.info("All data in data directory was removed")

ec_index = None

def parse_local_enzyme_dat():
  logger.info("Parsing local enzyme data file...")
  global ec_index
  ec_index = parse_enzyme_dat(ENZYMESDAT)

def setup():
  enable_logger("INFO")
  logger.info("Initializing emzyme data...")
  logger.info("Data directory: {}".format(APPDATADIR))
  logger.info("Downloading enzyme data...")
  APPDATADIR.mkdir(parents=True, exist_ok=True)
  downloaded = download(FTP_URL, ENZYMESDAT)
  if downloaded:
    logger.info("Downloaded enzyme data file.")
  else:
    logger.error("No enzyme data file found.")
    raise Exception("No enzyme data file found.")
  parse_local_enzyme_dat()

def __auto_setup():
  if not Path(ENZYMESDAT).exists():
    setup()
  else:
    parse_local_enzyme_dat()

def parse_enzyme_dat(enzyme_dat_filename):
    ec_index = {}
    ec_index["main"] = {}
    ec_index["alt"] = {}
    with open(enzyme_dat_filename) as f:
        current_id = None
        for line in f:
            line = line.rstrip()
            if line == "//":
                current_id = None
                continue
            line_type = line[0:2]
            if line_type == "ID":
                current_id = line[5:]
            elif line_type == "DE" and current_id is not None:
                name = line[5:].rstrip(".")
                if "Transferred entry" not in name:
                  ec_index["main"][name] = current_id
            elif line_type == "AN" and current_id is not None:
                name = line[5:].rstrip(".")
                if "Transferred entry" not in name:
                  ec_index["alt"][name] = current_id
    return ec_index

def search(enzyme_name, min_score=90, index=None, separator="\t"):
    if index is None:
      global ec_index
      index = ec_index
    main_names = index["main"]
    alternative_names = index["alt"]
    if enzyme_name in main_names:
        return "MAIN"+separator+main_names[enzyme_name]+separator+enzyme_name
    elif enzyme_name in alternative_names:
        return "ALT"+separator+alternative_names[enzyme_name]+\
            separator+enzyme_name
    else:
        match = process.extractOne(enzyme_name, main_names.keys())
        if match is not None and match[1] >= min_score:
          return "FUZZY:" + match[0] + separator + main_names[match[0]] +\
              separator + enzyme_name
        else:
          match = process.extractOne(enzyme_name, alternative_names.keys())
          if match is not None and match[1] >= min_score:
            return "FUZZY:" + match[0] + separator + \
                alternative_names[match[0]] +\
                separator + enzyme_name
          else:
            return "NOT_FOUND" + separator + separator + enzyme_name
